packages/pyright-internal/src/languageService/indentationUtils.ts

```diff
@@ -421,26 +418,364 @@ function _getLineEndingLength(parseResults: ParseResults, line: number) {
+                    };
+                }
+
+                // calculate indentation upto line offset given.
+                charOffset++;
+                indentation++;
+        }
+    }
+
+    return {
+        charOffset,
+        indentation,
+    };
+}
+
+export function reindentSpan(
+    parseResults: ParseResults,
+    span: TextRange,
+    indentation: number,
+    indentFirstToken = true
+) {
+    let indentDelta = 0;
+    const texts: string[] = [];
+
+    // Currently _convertTokenStreams converts text in the span as whitespace and non whitespace
+    // and then this function puts those back to string with reidentation if needed.
+    //
+    // Another approach we can take is converting the text in 2 chunks that require reindentation and not
+    // and process chunks that require reindentation line by line (like how it currently does for
+    // multiline doc comments) and put chunks that don't require reindentation as it is.
+    const tokenInfo = _convertTokenStreams(parseResults, span);
+    let previousInfo = tokenInfo[0];
+
+    indentDelta =
+        indentation -
+        _getIndentationFromText(parseResults, previousInfo.range.start.line, previousInfo.range.start.character)
+            .indentation;
+
+    if (previousInfo.multilineDocComment) {
+        texts.push(..._reindentLinesFromText(parseResults, previousInfo, indentDelta));
+    } else {
+        if (indentFirstToken) {
+            texts.push(_createIndentationString(parseResults, indentation));
+        }
+
+        texts.push(previousInfo.text);
+    }
+
+    for (let i = 1; i < tokenInfo.length; i++) {
+        const info = tokenInfo[i];
+        if (info.firstTokenOnLine) {
+            texts.push(
+                parseResults.tokenizerOutput.predominantEndOfLineSequence.repeat(
+                    info.range.start.line - previousInfo.range.end.line
+                )
+            );
+
+            if (info.multilineDocComment) {
+                texts.push(..._reindentLinesFromText(parseResults, info, indentDelta));
+            } else {
+                // Put indentation for the first token on the line.
+                texts.push(
+                    _createIndentationString(
+                        parseResults,
+                        Math.max(
+                            0,
+                            _getIndentationFromText(parseResults, info.range.start.line, info.range.start.character)
+                                .indentation + indentDelta
+                        )
+                    )
+                );
+                texts.push(info.text);
+            }
+        } else {
+            // Put whitespace between 2 tokens on same line
+            // token1[space]token2
+            texts.push(' '.repeat(info.range.start.character - previousInfo.range.end.character));
+            texts.push(info.text);
+        }
+
+        previousInfo = info;
+    }
+
+    return texts.join('');
+}
+
+interface TokenInfo extends TextRange {
```

Comment:
For consistency with the rest of the code base, please move interface definitions before functions in the file.

Reply:
okay